Sources/GRPC/Concurrency/AsyncWriter.swift

```diff
@@ -0,0 +1,266 @@
+  internal typealias Element = Delegate.Element
+  internal typealias End = Delegate.End
+
+  /// A value pending a write.
+  private struct Pending<Value> {
+    var value: Value
+    var continuation: UnsafeContinuation<Void, Error>
+  }
+
+  private typealias PendingElement = Pending<Element>
+  private typealias PendingEnd = Pending<End>
+
+  private enum CompletionState {
+    /// Finish hasn't been called yet. May move to `pending` or `completed`.
+    case incomplete
+    /// Finish has been called but the writer is paused. May move to `completed`.
+    case pending(PendingEnd)
+    /// The completion message has been sent to the delegate. This is a terminal state.
+    case completed
+
+    /// Move from `pending` to `completed` and return the `PendingCompletion`. Returns `nil` if
+    /// the state was not `pending`.
+    mutating func completeIfPending() -> PendingEnd? {
+      switch self {
+      case let .pending(pending):
+        self = .completed
+        return pending
+      case .incomplete, .completed:
+        return nil
+      }
+    }
+
+    var isPendingOrCompleted: Bool {
+      switch self {
+      case .incomplete:
+        return false
+      case .pending, .completed:
+        return true
+      }
+    }
+  }
+
+  /// A lock which must be help when accessing or modifying mutable state held by this storage.
+  private var lock: Lock
+
+  /// Elements and continuations which have been buffered but are awaiting consumption by the
+  /// delegate.
+  private var pendingElements: CircularBuffer<PendingElement>
+
+  /// The maximum number of pending elements. `pendingElements` must not grow beyond this limit.
+  private let maxPendingElements: Int
+
+  /// The completion state of the writer.
+  private var completionState: CompletionState
+
+  /// Whether the writer is paused.
+  private var isPaused: Bool = false
+
+  /// The delegate to process elements. By convention we call the delegate before resuming any
+  /// continuation.
+  private var delegate: Delegate
+
+  fileprivate init(maxPendingElements: Int, delegate: Delegate) {
+    self.lock = Lock()
+    self.maxPendingElements = maxPendingElements
+    self.pendingElements = CircularBuffer(initialCapacity: maxPendingElements)
+    self.completionState = .incomplete
+    self.delegate = delegate
+  }
+
+  deinit {
+    let completion = self.lock.withLock { self.completionState }
+    switch completion {
+    case .completed:
+      ()
+    case .incomplete, .pending:
+      assertionFailure("writer has not completed is pending completion")
+    }
+  }
+
+  fileprivate func pause() {
+    self.lock.withLockVoid {
+      self.isPaused = true
+    }
+  }
+
+  fileprivate func resume() {
+    self.lock.withLockVoid {
+      self.isPaused = false
+    }
+
+    while let nextWrite = self.nextWrite() {
+      switch nextWrite {
+      case let .element(pendingElement):
+        self.delegate.write(pendingElement.value)
+        pendingElement.continuation.resume()
```

Comment:
I'm not sure about this. Let's consider we have a buffer full with 16 writes waiting. After we have written the first element, do we want to allow more writes from the users side? There are still 15 writes pending... 

Further this is in a non detached loop. This means we will def. write all 16 items, and we will signal to everyone waiting, you can write again...

If this was an actor, we could consider using `await Task.yield` to check after a number of writes if the target is still writable and only in those cases to start succeed the continuations.